x-pack/plugins/apm/server/routes/profiling/route.ts

```diff
@@ -156,6 +202,163 @@ const profilingFunctionsRoute = createApmServerRoute({
+      await Promise.all([
+        core.elasticsearch.client,
+        await plugins.profilingDataAccess?.start(),
+        getApmEventClient(resources),
+      ]);
+    if (profilingDataAccessStart) {
+      const { serviceName } = params.path;
+      const {
+        start,
+        end,
+        kuery,
+        transactionName,
+        transactionType,
+        environment,
+      } = params.query;
+
+      // TODO: This will be used once the ES bug is fixed
+      const indices = apmEventClient.getIndexFromProcessorEvent(
+        ProcessorEvent.transaction
+      );
+      console.log('### caue  handler:  index:', indices);
+
+      return await profilingDataAccessStart?.services.fetchFlamechartData({
+        core,
+        esClient: esClient.asCurrentUser,
+        indices: ['traces-apm-default'],
+        stacktraceIdsField: TRANSACTION_PROFILER_STACK_TRACE_IDS,
+        totalSeconds: end - start,
+        query: {
+          bool: {
+            filter: [
+              ...kqlQuery(kuery),
+              ...termQuery(SERVICE_NAME, serviceName),
+              ...termQuery(TRANSACTION_NAME, transactionName),
+              ...environmentQuery(environment),
+              ...termQuery(TRANSACTION_TYPE, transactionType),
+              {
+                range: {
+                  ['@timestamp']: {
+                    gte: String(start),
+                    lt: String(end),
+                    format: 'epoch_second',
+                  },
+                },
+              },
+            ],
+          },
+        },
+      });
+    }
+
+    return undefined;
+  },
+});
+
+const transactionsFunctionsRoute = createApmServerRoute({
+  endpoint: 'GET /internal/apm/services/{serviceName}/transactions/functions',
+  params: t.type({
+    path: t.type({ serviceName: t.string }),
+    query: t.intersection([
+      environmentRt,
+      t.type({
+        start: isoToEpochSecsRt,
+        end: isoToEpochSecsRt,
+        startIndex: toNumberRt,
+        endIndex: toNumberRt,
+        transactionName: t.string,
+        transactionType: t.string,
+      }),
+      kueryRt,
+    ]),
+  }),
+  options: { tags: ['access:apm'] },
+  handler: async (resources): Promise<TopNFunctions | undefined> => {
+    const { context, plugins, params } = resources;
+    const core = await context.core;
+
+    const [esClient, profilingDataAccessStart, apmEventClient] =
+      await Promise.all([
+        core.elasticsearch.client,
+        await plugins.profilingDataAccess?.start(),
+        getApmEventClient(resources),
+      ]);
+    if (profilingDataAccessStart) {
+      const {
+        start,
+        end,
+        startIndex,
+        endIndex,
+        kuery,
+        transactionName,
+        transactionType,
+        environment,
+      } = params.query;
+      const { serviceName } = params.path;
+
+      // TODO: This will be used once the ES bug is fixed
+      const indices = apmEventClient.getIndexFromProcessorEvent(
```

Comment:
What is the bug?

Reply:
This: https://github.com/elastic/elasticsearch/pull/105372